### PR TITLE
Added support for passing in Report Parameters

### LINKED ISSUE #97: New-RsSubscription does not accept -Parameters argument (and sets it to null)
**Do you want to request a *feature* or report a *bug*?**
*feature*

**What is the current behavior?**
$Parameters is hardcoded to null and there is no API hook exposed to override this hardcoded behavior. See: https://github.com/Microsoft/ReportingServicesTools/blob/master/ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1#L205

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**
I don't think I really need to explain this.

**What is the expected behavior?**
Feature parity with CreateSubscription SSRS WSDL endpoint.

**Which versions of Powershell and which OS are affected by this issue? Did this work in previous versions of our scripts?**

```powershell
PS C:\> $PSVersionTable

Name                           Value                                                                                                                                                                                                 
----                           -----                                                                                                                                                                                                 
PSVersion                      5.1.15063.483                                                                                                                                                                                         
PSEdition                      Desktop                                                                                                                                                                                               
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                                                                               
BuildVersion                   10.0.15063.483                                                                                                                                                                                        
CLRVersion                     4.0.30319.42000                                                                                                                                                                                       
WSManStackVersion              3.0                                                                                                                                                                                                   
PSRemotingProtocolVersion      2.3                                                                                                                                                                                                   
SerializationVersion           1.1.0.1                                                                                                                                                                                               
```

### PR DESCRIPTION
**Pull request by jzabroski**
Fixes #97 .

Tested on SQL Server 2016, PowerShell 5, Windows Server 2016

### CODE REVIEW

**Comment by jtarquino** [comment]
>             Report server proxy to use. Use "New-RsWebServiceProxy" to generate a proxy object for reuse. Useful when repeatedly having to connect to multiple different Report Server.

Please add the Parameters parameter to the documentation 

---
Refers to: ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1:23 in 64384cd. [](commit_id = 64384cdc46c9957c1b4edaf3f3b843df1ce98551, deletion_comment = False)

**Comment by jtarquino** [comment]
>             This command will establish a connection to the Report Server located at http://myserver/ReportServer using current user's credentials and create a subscription based on other parameters specified.

You might want to add a example of the usage of the new parameter

---
Refers to: ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1:136 in 64384cd. [](commit_id = 64384cdc46c9957c1b4edaf3f3b843df1ce98551, deletion_comment = False)

File: ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1 [hunk 1]
```diff
@@ -313,7 +371,9 @@ function New-RsSubscription
         }
         catch
         {
-            throw (New-Object System.Exception("Exception occurred while creating subscription! $($_.Exception.Message)", $_.Exception))
+            $ex = $_.Exception;
+            $ErrorMessage = $ex.Message;
+            throw (New-Object System.Exception("Exception occurred while creating subscription! ErrorMessage: [$ErrorMessage]", $Exception))
```
Comment (jtarquino) [inline ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1]: not mandatory , but it would be great if tests for this new functionality are added, the coverage there is already pretty low

Comment (jzabroski) [reply ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1]: I might be able to add tests next week but it already took me a long time to fix everything.

Can you update the README.md with instructions on how to set-up a brand new PC with SSRS using the ReportingServicesTools?  As an outsider, it was very hard even figuring out the right scaffolding steps to deploy Reports, much less configure Report Server. There is strange vocabulary in the ReportingServicesTools Admin synposes, like references to "PBIRS," which I don't use and that intimidates me.

Comment (jtarquino) [reply ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1]: The most up to date sample on how to configure it using the powershell tools is referencing the continuous integration script , where it uses the powershell tools to configure the test machine https://github.com/Microsoft/ReportingServicesTools/blob/d06abfc41ae76c51eb548574972498760afe9605/appveyor.yml#L66 

The original idea of the admin cmdlets were created mostly to enable automation, feel free to add more samples or propose changes to make it more usable for all the users.

And for PBIRS it stands for Power BI Report Server which is a superset of Reporting Services (https://docs.microsoft.com/en-us/power-bi/report-server/get-started)

**Comment by jzabroski** [comment]
Excellent, thank you.

File: ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1 (hunk 1 above)
Comment (jzabroski) [reply ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1]: @jtarquino  I think the part that confused me is in a lot of the Admin powershell functions it refers to PBIRS instead of just the subset of Reporting Services I might colloquially call "SSRS". What do you think about clarifying that part of the documentation?

Comment (bateks) [reply ReportingServicesTools/Functions/CatalogItems/New-RsSubscription.ps1]: So, are these enhancements already applied to the current version of ReportingServicesTools?
If they are not: what are we waiting for? I really need them and already tested\used in my environment, but "clumsy" way.

**Comment by Periscope34326** [comment]
Testing on SQL Server 2012, PowerShell 5, Windows Server 2016. Running into a credential issue here when trying to create a new subscription that writes to a fileshare. This seems to arise whether I use default credentials or try to supply alternate credentials. All the parameters I am using appear to be properly formatted etc.

I get:
> Microsoft.ReportingServices.Diagnostics.Utilities.InvalidExtensionParameter: One of the extension parameters is not valid for the following reason: The setting DEFAULTCREDENTIALS is unknown."]


I also note that extension parameters seem to be case sensitive, worth noting in documentation

**Comment by Periscope34326** [comment]
Testing on SQL Server 2012, PowerShell 5, Windows Server 2016, creating new **emailed** subscriptions works perfectly for me

**Comment by jzabroski** [comment]
@Periscope34326  Were you using my hotfix or the current version of ReportingServicesTools?

**Comment by jzabroski** [comment]
I've never written tests for these because I was told to go figure out how to set-up a test environment in ReportingServicesTools using the build pipeline script Appveyor.yml , and when I tried to actually set-up a test environment based on that, it failed, so I gave up and _mostly_* forgot about it. On top of that, there are features in ReportingServicesTools I have _never_ used that would require _lots of effort_ to learn what they are (Power BI Report Server or PBIRS) just to run the test suite as well as fix any issues with the Appveyor.yml file.

I say _mostly_ forgot about it, because I started looking into how to run SSRS inside a docker container, because the Appveyor.yml script is rather messy compared with most build scripts I maintain: 

- It uses random download URLs and strings to exe's on those URL endpoints
- It doesn't use a package manager like Chocolatey to get those dependencies and explicitly list what version number it is testing.  If my code succeeds, and the build server fails, I have no transparency into why.  "Works on my machine". 
- When I started using this project, PBIRS didn't even exist as part of the API
- Parts of the test suite depend upon the system being in a very specific state, which is right after a clean install

Sorry all.

**To be clear, the hold up is writing test coverage for this code. And there is an undefined amount of time required to figure out how to create a test environment for SSRS, due to all my concerns above**

Also, as best as I can tell, Microsoft has rolled resources off this project.  I haven't spoken to anyone at Microsoft, but the vibe I get is they're pushing the cloud-based solutions and SSRS is a desktop solution that does not fit with Satya Nadella's focus on the cloud. https://kohera.be/blog/azure-cloud/opinion-is-ssrs-on-azure-happening/

**Comment by jtarquino** [comment]
I think there is a missunderstanding here, you can use the same environment where you developed the changes in order to add the tests, there is a section that details how to do local testing and development https://github.com/Microsoft/ReportingServicesTools/#local-testing-and-development , if you want to execute only your new test you can even run that one with the same type of command that run the suite https://github.com/Microsoft/ReportingServicesTools/blob/master/Test.ps1

**Comment by jzabroski** [comment]
>I think there is a missunderstanding here

I disagree.  It's more likely I have a deeper understanding of the status of this than anyone else, since I actually worked on it and fixed the issue.  I just haven't been able to run the test suite.

> there is a section that details how to do local testing and development https://github.com/Microsoft/ReportingServicesTools/#local-testing-and-development 

Then it is not detailed enough, because I followed those instructions and had errors running the test suite. - You need _everything installed_ and you _also_ need specific versions of ReportServer.exe installed _or the test suite will fail_.

Recall, I asked for an updated README.md, and [you replied](https://github.com/Microsoft/ReportingServicesTools/pull/186#discussion_r210652898):

> The most up to date sample on how to configure it using the powershell tools is referencing the continuous integration script , where it uses the powershell tools to configure the test machine

The test suite failed because I don't use PBIRS, and don't have any desire to learn it, since I will never be compensated for that time.  We will _never_ adopt this technology.  Maybe I was _really close_ to finishing this and gave up, but I gave up with _all the red flags I raised above_ as my reasons for giving up.  You just have to make educated guesses as to when and where to cut losses.

**Comment by jtarquino** [comment]
I understand your decision and I'm sympathetic and thankful for the time that require any contribution, each one of them with its own learning curve.

For future reference I would like to provide additional ideas an options to consider for anyone interested in add the test coverage required for new features

Indeed there are certain tests that depends in a specific version of PBIRS/SSRS but as far I remember (and my memory could be failing me) you don't need it for the are you are modifying
I think you can run only the test of the function you are modifying, checking the code there is a environment variable [$env:PesterServerUrl](https://github.com/Microsoft/ReportingServicesTools/blob/ffe49b6f1e7dde68503ce94a3ffc87b76d9341c5/Tests/CatalogItems/New-RsSubscription.Tests.ps1#L4
) that you can set to your report server url (which is http://yourservername/reportserver) 

Also is not the most optimal way because the time that takes to execute, but you could write the test in your PR and just submit , the CI system will run your new test for you

**Comment by jzabroski** [comment]
It's cool, I'm easy.  Just trying to lay out the _why_ and move on.  Feel I owe everyone at least that much.

**Comment by gperrego** [comment]
@jtarquino can you write the tests so the community can get the functionality?  We need it

**Comment by JamesDBartlett** [comment]
@jtarquino I'm also curious about the next steps for this to get merged. I don't have any background in unit testing, so I'm not sure how much help I can be, but I have a looming deadline for migrating SSRS subscriptions to a new server from an older one before it gets retired, so if there's anything I can do to help move this along, please let me know.

**Comment by JamesDBartlett** [comment]
@jzabroski First of all, thank you for trying so hard to get this working, and it really bums me out that MS has dropped the ball on getting your bugfixes integrated into their product.

Does the code in your fork work as-is? I have a looming deadline for migrating a few hundred SSRS report subscriptions, and given that the alternative is copy-pasting every field manually over the next several weeks, let's just say I'm not terribly concerned about test coverage at the moment. ;-)

**Comment by jzabroski** [comment]
Yes, it works.

Assuming you are working with the raw XML from your previous SSRS database, you can write the following PowerShell snippet to massage the XML into a PS hashtable:

```powershell
    $parametersHashTable = @{};

    if ($null -ne $parameters.ParameterValues.ParameterValue)
    {
        $parameters.ParameterValues.ParameterValue | ForEach-Object { $parametersHashTable.Add($_.Name, $_.Value) }
    }
    else
    {
        $parametersHashTable = $null;
    }
```
You then pass `$parametersHashTable` into `-Parameters $parametersHashTable` when calling `New-RsSubscription`.  The massaging is required due to a limitation/breaking change in the SOAP API between SSRS 2005 and SSRS 2014.  I don't remember which version of SSRS broke the API, but handling of blank and null values became conflated at some point and they never fixed it.  IIRC, this may be the scenario when your Reports database has a value of `<ParameterValues />` in the dbo.Subscriptions.Parameters column.

I recommend calling `New-RsSubscription` with `-Verbose` so you can see my diagnostics.

### Soap Box
To be honest, I think SSRS as it exists today is a dead product.  They already screwed over a lot of customers by removing multi-tenancy and not explaining how to use Docker to circumvent the single-tenant limit.  I'm told by people who do reporting for a living that Microsoft has effectively stopped giving away SSRS for free and the removal of multi-tenancy is so that you'll buy "Azure SQL Managed Instances" and "PowerBI Premium".  But, don't take non-Microsoft employee word for it.  Look at their documentation, here: https://docs.microsoft.com/en-us/azure/sql-database/sql-database-managed-instance-index where they clearly state: "Move your SSIS packages to Azure Data Factory, SSAS tabular models to Azure Analysis Service, and SSRS reports to PowerBI Premium."

So, to recap, rather than provide tutorials to customers on how they can deploy SSRS via Docker to save costs, the messaging is:
* Buy Azure SQL Managed Instances
* Buy PowerBI Premium

I'd suggest you direct your complaintsto PowerBI Principal Program Manager Christopher Finlan: https://powerbi.microsoft.com/en-us/blog/whats-new-for-sql-server-2019-reporting-services-rc1/ - maybe you work for a large enough organization where your organization's opinion on this matters.

**Comment by JamesDBartlett** [comment]
@jzabroski Wow, thank you so much for that detailed reply! I'm wrapping up for the day right now, but I will be putting your suggestion into practice first thing tomorrow, and I'll let you know how it goes. 

I started this job a little under a year ago, and it was explained to me on my first day that we had already begun a gradual transition from SSRS to Power BI, because the key decision makers had already seen the writing on the wall for SSRS, and our users were asking for more dynamic reporting anyway. 

Fast forward to today. We still have thousands of active SSRS reports which haven't been replicated in Power BI yet, but the servers they live on are being retired, so we have to migrate and continue supporting them for now, until our Power BI offerings reach parity with our existing SSRS portfolio, and our various app dev teams have a chance to re-tool and refactor their code for integration with Power BI. It'll take a while, but I think it'll be the right move for us in the long term.

### Reply Soap Box
I agree with your assessment that SSRS is a dead product. It's frustrating to see support for it being killed off so suddenly, when there are still **so many** valid use cases for it out there, and probably millions of users that rely on SSRS reports to help them make business decisions every day. I know Microsoft has a lot of irons in the fire, and I do think that Power BI will eventually be able to replace SSRS entirely, but it's not there yet. Most Microsoft products get a pretty generous sunset period, with an EOL date announced **long before** they discontinue maintenance and support, so it's very disappointing to see them pull the rug out from under such a mature, widely-used product without giving their customers that usual courtesy (or even so much as announcing that they are doing it at all), and before the product slated to take its place is truly ready to do so. 

Sorry for writing a novel. It seems we both have strong feelings about this issue. :-) 
I hope you have a great night!

~ James

**Comment by gperrego** [comment]
@jtarquino Ok, whats the deal, can this be accepted and merged into Main?  We can write the unit tests if that is the show stopper but this is killing me.  We have workable code and its just rotting on the shelf.

**Comment by jtarquino** [comment]
@gperrego, yes the stopper is the unit testing which ensures that new changes will not break fundamentally existing cases and also validate that the scenario works properly in the happy paths
For the latter as I see from this thread is working fine and is not breaking any existing scenario so I feel comfortable merging it, however it will be great if anyone add some unit tests for it

**Comment by jzabroski** [comment]
@jtarquino  Thank you.
